json-everything.net/wwwroot/md/vocabs-data-2022.md

```diff
@@ -0,0 +1,123 @@
+
+## 3. The `data` Keyword
+
+### 3.1 Syntax and Semantics
+
+The value of `data` must be an object.  The keys of the object correspond to valid JSON Schema keywords, and the values MUST be one of
+
+- JSON Pointers per [RFC 6901](https://www.rfc-editor.org/rfc/rfc6901)
+- [Relative JSON Pointers](https://json-schema.org/draft/2019-09/relative-json-pointer.html)
+- URI references per [RFC 3986](https://datatracker.ietf.org/doc/html/rfc3986), potentially with JSON Pointer fragments
+
+The set of keywords in the schema or subschema containing `data` SHOULD be distinct from the set of keys defined within `data`.  Behavior in the event of an overlap is not defined.
+
+`data` operates in two phases:
+
+1. All of the values are dereferenced per sections 3.2 and 3.3.
+2. The keywords and values then form a schema which is applied to the instance at the current location.
+
+The validation and annotation results of `data` are those of the formed schema.  More detail regarding output can be found in section 3.4.
+
+### 3.2 Contextual Behavior
+
+`data` MUST be processed contextually in accordance with the draft of the schema in which it is used.  For example, if `data` is used in a schema that declares draft 2019-09, then
+
+- the keys SHOULD be recognized by draft 2019-09 or defined in other referenced vocabularies, and
+- the URI references MUST be resolved following the same rules as `$ref` as defined by draft 2019-09.
+
+If a key is not a recognized keyword in the schema's draft (e.g. `$dynamicAnchor` in a draft 2019-09 schema), then this key is ignored just as it would be ignored if it were in the schema or subschema containing `data`.
+
+### 3.3 URI Resolution
+
+The values of `data` are dereferenced in different ways depending on the format of the value.
+
+If the value is a JSON Pointer, it is resolved against the instance root.
+
+If the value is a Relative JSON Pointer, it is resolved against the instance at the location currently being evaluated.
+
+If the value is a URI reference, it must be resolved in accordance with the rules of `$ref` resolution for the containing spec, as per section 3.2.  The primary different between resolving URIs for `data` and resolving URIs for `$ref` is that `$ref` requires that the indicated data must represent a valid schema object whereas `data` can reference any valid JSON value.
+
+External document data can be referenced using an absolute URI base identifier along with a JSON Pointer fragment.
+
+Implementations SHOULD be able to download external JSON documents, but MAY provide means to pre-load and cache such documents against the identifiers by which they will be referenced.  Documents downloaded from URIs which contain a JSON Pointer fragment MUST be interpreted using a media type that allows resolution of such fragments.
```

Comment:
> So you definitely want to restrict this to JSON documents?

Not necessarily.  You're right in that it needs to be mappable to JSON.  However, I don't know that I can realistically support non-JSON, much less "any mappable format."  That's really open-ended and not practical.

Reply:
This is why I was pushing for an emphasis on pre-populating a URI-to-data storage (like the schema storage), because if you want to go outside of HTTP and JSON, that can be done by an app that fetches, parses, and maps the data into the data model, and then `data` just looks it up by URI.

If `data` has to do the fetching, then we're pretty much limited to a small set of URI schemes/protocols and media types.

If the pre-load behavior becomes the recommended default, and the auto-fetch an optional extra, then limiting the auto-fetch to HTTP(S) and JSON-compatible media types is fine, as it doesn't constrain the overall usage.

I'm particularly thinking of a usage of `data` to access values from a database that is not served over HTTP.